File: src/algorithms/arrays/two-number-sum/solution3.js
// STRATEGY:
// Sort array - traverse from both ends with left/right pointers
// COMPLEXITY:
// O(nlog(n)) time | O(1) space

function twoNumberSum(array, targetSum) {
    array.sort((a, b) => a - b);
    let left = 0;
    let right = array.length - 1;

    while (left < right) {
        const currentSum = array[left] + array[right];

        if (currentSum === targetSum) {
            return [array[left], array[right]];
        } else if (currentSum < targetSum) {
            left++;
        } else if (currentSum > targetSum) {
            right--;
        };
    };

    return [];
};

exports.twoNumberSum = twoNumberSum;

File: src/algorithms/arrays/two-number-sum/solution3.go
package main

import "sort"

// O(nlog(n)) time | O(1) space
func TwoNumberSum(array []int, target int) []int {
	sort.Ints(array)
	left, right := 0, len(array)-1
	for left < right {
		currentSum := array[left] + array[right]
		if currentSum == target {
			return []int{array[left], array[right]}
		} else if currentSum < target {
			left += 1
		} else {
			right -= 1
		}
	}
	return []int{}
}